### PR TITLE
Tests and fixes for the ESDT integration

### PR DESCRIPTION
**Pull request by camilbancioiu**
Add tests and fixes for the ESDT API in smart contracts.

Two integration tests have been added, which use a Rust contract (with the new API). These tests validate the getter API (`getESDTValue()` and `getESDTTokenName()`):
* the first test uses an artificial direct call with injected ESDT tokens, which the contract must recognize
* the second test is called indirectly, via an `ExecuteOnDestContext()` call to the built-in function `ESDTTransfer`, with the contract call appended to the arguments

More comprehensive tests will be added in the future.

Also added finishing changes for an older issue: in-shard protection for sync-exclusive APIs like `executeOnSameContext()` and `executeOnDestContext()`.

Certain tests were adjusted to expect less gas to be consumed by test contracts, due to a minor C optimization.

### CODE REVIEW

File: arwen/elrondapi/elrondei.go [hunk 1]
```diff
@@ -1418,25 +1422,25 @@ func executeOnSameContext(
 	gasToUse = metering.GasSchedule().BaseOperationCost.DataCopyPerByte * uint64(actualLen)
 	metering.UseGas(gasToUse)
 
-	if arwen.WithFault(err, context, false) {
+	if err != nil {
```
Comment (sasurobert) [inline arwen/elrondapi/elrondei.go:1425]: arwen.WithFault must remain.

File: arwen/elrondapi/elrondei.go [hunk 2]
```diff
@@ -1395,14 +1395,18 @@ func executeOnSameContext(
 	gasToUse := metering.GasSchedule().ElrondAPICost.ExecuteOnSameContext
 	metering.UseGas(gasToUse)
 
-	send := runtime.GetSCAddress()
-	dest, err := runtime.MemLoad(addressOffset, arwen.AddressLen)
-	if arwen.WithFault(err, context, false) {
+	sender := runtime.GetSCAddress()
+	destination, err := runtime.MemLoad(addressOffset, arwen.AddressLen)
+	if err != nil {
```
Comment (sasurobert) [inline arwen/elrondapi/elrondei.go:1400]: arwen.WithFault must remain.

Comment (sasurobert) [reply arwen/elrondapi/elrondei.go:1400]: here the execution should stop

File: arwen/elrondapi/elrondei.go [hunk 3]
```diff
@@ -1395,14 +1395,18 @@ func executeOnSameContext(
...
+		return 1
+	}
+
+	if !areInSameShard(host, sender, destination) {
```
Comment (sasurobert) [inline arwen/elrondapi/elrondei.go:1404]: arwen.WithFault must remain.

File: arwen/elrondapi/elrondei.go [hunk 4]
```diff
@@ -1657,24 +1673,23 @@ func executeReadOnly(
 	gasToUse = metering.GasSchedule().BaseOperationCost.DataCopyPerByte * uint64(actualLen)
 	metering.UseGas(gasToUse)
 
-	if arwen.WithFault(err, context, runtime.ElrondAPIErrorShouldFailExecution()) {
+	if err != nil {
```
Comment (sasurobert) [inline arwen/elrondapi/elrondei.go:1676]: arwen.WithFault must remain.  please make the changes back. the execution must stop in these errors. We want to have the protected arwen, in which any error really stops the execution.

File: arwen/host/execution_test.go [hunk 5]
```diff
@@ -1140,53 +1060,6 @@ func TestExecution_AsyncCall_CallBackFails(t *testing.T) {
 	require.Equal(t, expectedVMOutput, vmOutput)
 }
 
-func TestExecution_AsyncCall_BuiltinFails(t *testing.T) {
```
Comment (sasurobert) [inline arwen/host/execution_test.go:1143]: why did you delete these tests ?

File: test/contracts/exchange/elrond.json [hunk 6]
```diff
@@ -0,0 +1,3 @@
+{
+    "language": "rust"
+}
```
Comment (sasurobert) [inline test/contracts/exchange/elrond.json:3]: add empty line

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: arwen/host/execution_test.go (hunk 5 above)
Comment (camilbancioiu) [reply arwen/host/execution_test.go:1143]: They were moved into `execution_builtin_test.go`.

File: arwen/elrondapi/elrondei.go (hunk 1 above)
Comment (camilbancioiu) [reply arwen/elrondapi/elrondei.go:1425]: Will restore them.

Comment (camilbancioiu) [reply arwen/elrondapi/elrondei.go]: Restored.

File: arwen/elrondapi/elrondei.go (hunk 2 above)
Comment (camilbancioiu) [reply arwen/elrondapi/elrondei.go]: Restored.

File: arwen/elrondapi/elrondei.go (hunk 3 above)
Comment (camilbancioiu) [reply arwen/elrondapi/elrondei.go:1404]: Restored.

File: arwen/elrondapi/elrondei.go (hunk 4 above)
Comment (camilbancioiu) [reply arwen/elrondapi/elrondei.go]: Restored.

File: test/contracts/exchange/elrond.json (hunk 6 above)
Comment (camilbancioiu) [reply test/contracts/exchange/elrond.json]: Added.

File: arwen/host/execution_test.go [hunk 7]
```diff
@@ -476,6 +476,10 @@ func TestExecution_ExecuteOnSameContext_Prepare(t *testing.T) {
 }
 
 func TestExecution_ExecuteOnSameContext_Wrong(t *testing.T) {
+	// CustomGasSchedule = config.MakeGasMapForTests()
```
Comment (sasurobert) [inline arwen/host/execution_test.go:479]: delete commented code

**Review by sasurobert — APPROVED** [review]

File: arwen/host/execution_test.go (hunk 7 above)
Comment (camilbancioiu) [reply arwen/host/execution_test.go]: Deleted.

**Review by sasurobert — APPROVED** [review]

File: arwen/contexts/runtime.go [hunk 8]
```diff
@@ -210,6 +212,15 @@ func (context *runtimeContext) SetVMInput(vmInput *vmcommon.VMInput) {
 		copy(context.vmInput.CallerAddr, vmInput.CallerAddr)
 	}
 
+	if vmInput.ESDTValue != nil {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply arwen/contexts/runtime.go:215]: No, only together

[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/contexts/runtime.go:215]: Will not add the check for now, due to protection from the protocol.

**Review by AdoAdoAdo — APPROVED** [review]
